packages/dds/tree/src/forest/types/reader.ts

```diff
@@ -0,0 +1,25 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Serializable } from "@fluidframework/datastore-definitions";
+import { TreeKey, TreeType } from "../..";
+
+export const enum TreeReadResult {
+    NotFound = -1,
+    Pending = 0,
+    Ok = 1,
+}
+
+export interface INodeReader {
+    type: TreeType;
+    keys: Iterable<TreeKey>;
+    length(key: TreeKey): number;
+    value: undefined | Serializable;
```

Comment:
this type was going to have a way to start iterating from it, right? You called it "fork?". I think we want that.''

Name wise, having `IterableTree.iterate(): TreeIterator` seems elegant.

Reply:
I did have a 'fork()', and agree we'll want that functionality.  Here's the interfaces as they appear in my draft repo.

```ts
export interface INodeReader {
    id: TreeId;
    type: TreeType;
    keys: Iterable<TreeKey>;
    length(key: TreeKey): number;
    value: undefined | Serializable;
    
    fork(): ITreeReader;
}

export interface ITreeReader extends INodeReader {
    push(key: TreeKey, index: number): TreeReadResult;
    pushFirst(key: TreeKey): TreeReadResult;
    pushLast(key: TreeKey): TreeReadResult;
    next(): TreeReadResult;
    prev(): TreeReadResult;
    pop(): TreeReadResult;
    
    visit((node: INodeReader) => TreeVisitResult): TreeVisitResult;
}
```

In this iteration, I've only pulled in the members I've implemented.